lib/rules.js

```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
+        return inv ? 0 : num
+    }
+
+    // Source: https://rosettacode.org/wiki/Nth_root#JavaScript
+    const n = root
+    const prec = precision
+
+    let x = 1 // Initial guess
+    for (let i = 0 ; i < prec ; i++) {
+        x = 1 / n * ((n - 1) * x + (num / Math.pow(x, n - 1)))
+    }
+
+    return inv ? 1 / x : x
+}
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+    (node) => {
+        const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+        if (isFactorable) {
+            const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+            return print(result.args[0]) === print(result.args[1])
+                ? {node} : null
+        }
+    },
+
+    (node) => {
+        const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+        return build.apply('pow', [result.args[0], build.numberNode(2)])
+    }
+)
+
+function getFactors(num){
+    var factors = []
+
+    for(var i = 1; i <= Math.sqrt(num); i++){
+        if(Number.isInteger(num/i)){
+            factors.push([i, num/i])
+        }
+    }
+
+    return factors
+}
+
+// TODO: handle case when exponent is a polynomial (x^b^4)
+// e.g. 12x^2 + 17x + 6 -> (3 x^1 + 2) (4 x^1 + 3)
+export const FACTOR_SUM_PRODUCT_RULE = defineRule(
+    (node) => {
+        if (query.isAdd(node) && node.args.length === 3) {
+            const [firstTerm, secondTerm, thirdTerm] = node.args
+
+            // First two terms should be polynomials
+            // Third term is either number or polynomial
+            const isTrinomial =
+                  isPolynomialTerm(firstTerm)
+                  && isPolynomialTerm(secondTerm)
+                  && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
+
+            if (isTrinomial) {
+                let isFactorable = false
+                const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+                const variables = Object.keys(coefficientMap).map(key => parse(key))
+                const coefficients = Object.keys(coefficientMap).map(key => {return coefficientMap[key]})
+
+                const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+                const thirdPoly =
+                      variables[2]
+                      ? query.isMul(variables[2])
+                      ? variables[2].args
+                      : [variables[2]]
+                      : null
+
+                const secondCoef = coefficients[1][0]
+                // Second polynomial to be matched
+                const firstMiddle = firstPoly.map(term => {
+                    const identifier = term.args[0]
+                    const exponent = build.number(query.getValue(term.args[1]) / 2)
+                    return build.applyNode('pow', [identifier, exponent])
+                })
+
+                let secondMiddle
```

Comment:
Can you provide an explanation of what `firstMiddle` and `secondMiddle` are and what the algorithm is at a high level?